7a_temp_coop_munge/src/data_parsers/parse_mo_usace.R

```diff
@@ -0,0 +1,427 @@
+  })
+}
+
+#' Coerce numeric columns
+#'
+#' This function converts user-specified columns to `numeric`. It
+#' assumes that the `Site` column is the only `chr` column.
+#'
+#' @param ls_df a list of `data.frames` generated by reading in excel sheets
+#'
+coerce_numeric_columns <- function(ls_df) {
+  lapply(ls_df, function(df){
+    # ID all the numeric columns
+    num_cols <- c(1:ncol(df)) %>%
+      .[!. %in% grep("site", names(df))] # find and remove site column
+
+    df[num_cols] <- sapply(df[num_cols], as.numeric)
+
+    return(df)
+  })
+}
+
+#' Validate column names
+#'
+#' MO USACE datasets have a wide variety of column names. This function
+#' checks for inconsistencies and fixes them. this function calls two
+#' helper functions: `run_name_checks` and `find_col_idx`
+#'
+#' @param ls_df a list of `data.frames` generated by reading in excel sheets
+#'
+validate_col_names <- function(ls_df) {
+
+  df_names <- names(ls_df)
+
+  mapply(run_name_checks, ls_df, site_pattern = df_names, SIMPLIFY = FALSE)
+}
+
+#' Column check logic
+#'
+#' Validate column names and rename to a consistent set of
+#' hard-coded names `c('site', 'date', 'temp', 'depth')`
+#'
+#' @param df data.frame
+#' @param site_pattern chr, partial match that should be use to verify if a column is the `site` column.
+#'
+run_name_checks <- function(df, site_pattern) {
+
+  # grab data.frame names for validation
+  chk_names <- names(df)
+
+  # check for the use of "sample id" instead of "site" and reset `chk_names`
+  if(!is.na(pmatch('samp', chk_names))){
+    names(df)[pmatch('samp', chk_names)] <- 'site'
+    chk_names <- names(df)
+    }
+
+  # check for "site" column names using a partial match
+  if(is.na(pmatch('site', chk_names))) {
+    #looks for the name of the data.frame which is typically a two letter abbreviation (e.g. "RA")
+    site_col_idx <- find_col_idx(df, site_pattern)
+    names(df)[site_col_idx] <- 'site'
+  } else {
+    names(df)[pmatch('sit', names(df))] <- 'site'
+  }
+
+  # these other should be there, so just making a stop right now
+  # check for "date" column names using a partial match
+  if(is.na(pmatch('dat', chk_names))) {
+    stop("Date column not found")
+  } else {
+    names(df)[pmatch('dat', names(df))] <- 'date'
+  }
+
+  # check for "depth" column names using a partial match
+  if(is.na(pmatch('dep', chk_names))) {
+    stop("Depth column not found")
+  } else {
+    names(df)[pmatch('dep', names(df))] <- 'depth'
+  }
+
+  # check for temp column names using a partial match
+  if(is.na(pmatch('tem', chk_names))) {
+    stop("Temp column not found")
+  } else {
+    names(df)[pmatch('tem', names(df))] <- 'temp'
+  }
+
+  return(df)
+}
+
+#' Identify column index
+#'
+#' Given a pattern, identify the column position of a column within a
+#' data.frame.
+#'
+#' @param df a data.frame
+#' @param pattern chr, pattern to match. Does not support partial matching.
+#'
+find_col_idx <- function(df, pattern) {
```

Comment:
I'd digging the modular code, but some of these really small functions may take things further than needed, esp the ones that are one line, like this. I think this one is used once and could appear here:

```r
site_col_idx <- find_col_idx(df, site_pattern)
names(df)[site_col_idx] <- 'site'
```
with 
```r
 site_col_idx <- which(apply(df, 2, function(x) any(grepl(site_pattern, x))))
 names(df)[site_col_idx] <- 'site'
 ```
 Instead, unless this function would be expected to be called elsewhere. 

Regardless, I found this one a little confusing, since I thought it was looking for names matches for column names, but it is actually looking at all of the values and returning TRUE (and then the index) for the columns that have these values, with partial matching (even though the doxy says it doesn't support partials)

```r
df[1:2, ]
# A tibble: 2 × 14
  ...1  date   time  `depth (m)` temp  conductivity       ph    turbidity          d.o.  d.o   `secchi depth`
  <chr> <chr>  <chr> <chr>       <chr> <chr>              <chr> <chr>              <chr> <chr> <chr>         
1 NA    MMDDYY HHMM  meters      øC    æS/cm              Units NTU                mg/l  % Sa… meters        
2 HT-1A 42109  1130  0.4         12.71 299.10000000000002 8.51  1.1000000000000001 10.6  99.1  NA           
```
☝️ I don't think "HT-1A" should have a partial match to "HT", but it does:
```r
apply(df[1:2, ], 2, function(x) any(grepl("HT", x)))
          ...1           date           time      depth (m)           temp   conductivity             ph 
          TRUE          FALSE          FALSE          FALSE          FALSE          FALSE          FALSE 
     turbidity           d.o.            d.o   secchi depth lake elevation        outflow          notes 
         FALSE          FALSE          FALSE          FALSE          FALSE          FALSE          FALSE 
```

My suggestion would be either to remove the function and put that code in the spot where it was calling it, or to change the name of this function to be more specific and also modify the documentation.

Reply:
I think I am going to go with second option suggested:

> change the name of this function to be more specific and also modify the documentation

I don't have a good reason for why the doc on `find_col_idx` says it doesn't support partial matching. Especially considering that `run_name_checks` explicitly says that it does! 😬 

I know it's a tiny function, but I want to keep it because I am making _a lot_ of calls to the `apply` family of functions in this parser. I think the wrappers keeps it readable. Also, I definitely thought I was going to use this function more than once! 😄